### DISCUSSION TITLE
ConfigMap rabbitmq-server-conf k8s

### DISCUSSION THREAD

**Discussion by FaraSys** [rabbitmq/cluster-operator#1163] [Q&A] [answered]
Hi everyone!
I have 3 cluster k8s v1.24
In two cluster I install cluster-operator via command:
1. k create ns rabbitmq-system
1. kubectl apply -f https://github.com/rabbitmq/cluster-operator/releases/latest/download/cluster-operator.yml (v. 1.14.0)
2. k apply -f rabbitmq.yaml
And I change settings in  configMap rabbitmq-server-conf. via command:
k apply -f cm.yaml -n rabbitmq-system. It work fine.

On the third cluster today I install  cluster-operator via command:
1. k create ns rabbitmq-system
1. kubectl apply -f https://github.com/rabbitmq/cluster-operator/releases/latest/download/cluster-operator.yml (v. 2.0)
2. k apply -f rabbitmq.yaml
When i change configmap with my settings, configmap returns to default values.

Why is this happening?

my cm:
apiVersion: v1
kind: ConfigMap
metadata:
  name: rabbitmq-plugins-conf
  namespace: rabbitmq-system
data:
  enabled_plugins: '[rabbitmq_peer_discovery_k8s,rabbitmq_prometheus,rabbitmq_management].'

---
apiVersion: v1
data:
  advancedConfig: |
    [
        {rabbit, [
            {consumer_timeout, undefined}
        ]}
    ].
  operatorDefaults.conf: |
    cluster_formation.peer_discovery_backend             = rabbit_peer_discovery_k8s
    cluster_formation.k8s.host                           = kubernetes.default
    cluster_formation.k8s.address_type                   = hostname
    cluster_partition_handling                           = pause_minority
    queue_master_locator                                 = min-masters
    disk_free_limit.absolute                             = 2GB
    cluster_formation.randomized_startup_delay_range.min = 0
    cluster_formation.randomized_startup_delay_range.max = 60
    cluster_name                                         = rabbitmq
  userDefinedConfiguration.conf: |
    total_memory_available_override_value = 1258291200
    cluster_partition_handling            = pause_minority
    vm_memory_high_watermark_paging_ratio = 0.99
    disk_free_limit.relative              = 1.0
    collect_statistics_interval           = 10000
    consumer_timeout                      = 3600000
kind: ConfigMap
metadata:
  name: rabbitmq-server-conf
  namespace: rabbitmq-system

**Comment by Zerpet** [comment]
The Operator will try to reconcile all dependent objects (ConfigMap, StatefulSet, Service, etc.) from the information in `RabbitmqCluster.spec`. This means that any changes to a dependent object will be overriden, as you are observing.

To configure RabbitMQ, you can use `RabbitmqCluster.spec.rabbitmq.additionalConfig` (see [this link](https://rabbitmq.com/kubernetes/operator/using-operator.html#additional-config)) and `RabbitmqCluster.spec.rabbitmq.advancedConfig`. The Operator will include your configuration in a dedicated config file inside RabbitMQ Pods.

I'm surprised you don't observe the same behaviour in 1.14. I believe there has been no changes for quite a while. I suspect there might be an additional factor causing this e.g. paused reconcilliation via label, error in reconcile loop, to name a few, that may be causing different behaviour in your 1.14 cluster.

**Comment by FaraSys** [comment]
@Zerpet Thank you for you answer/
That's right. In 1.14.0, the configuration does not change after my changes, but in 2.0, my changes immediately return to their original values.

Honestly, I don't quite understand how to correctly specify these settings. I have downloaded cluster-operator.yaml, but how should I change it?
I changed like this but it doesn't work. Look my screen
![image](https://user-images.githubusercontent.com/67997008/196410957-2bb5ef1f-a50e-45d2-8c9f-5e0ee2bd9772.png)

**Comment by FaraSys** [comment]
After applying this yaml file, a standard configmap is created again
![image](https://user-images.githubusercontent.com/67997008/196411196-64349ba5-f0bb-481f-a764-424c288835b3.png)

**Comment by mkuratczyk** [comment, marked as answer]
Have a look at some of the [examples](https://github.com/rabbitmq/cluster-operator/blob/main/docs/examples/), especially [custom-config](https://github.com/rabbitmq/cluster-operator/blob/main/docs/examples/custom-configuration/rabbitmq.yaml).

**Comment by FaraSys** [comment]
@mkuratczyk Thank you. This solved my issue.
![image](https://user-images.githubusercontent.com/67997008/196601534-1795a2ec-0393-4f24-9fba-f752a26af7da.png)
